protos/consensus.proto

```diff
@@ -0,0 +1,325 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// -----------------------------------------------------------------------------
+
+syntax = "proto3";
+
+option java_multiple_files = true;
+option java_package = "sawtooth.sdk.protobuf";
+option go_package = "consensus_pb2";
+
+
+// Data Structures
+
+// Payload is the opaque payload to send to other nodes. The
+// interpretation of message_type is left to the Consensus Engine
+// implementation.
+message ConsensusMessage {
+  string message_type = 1;
+  bytes payload = 2;
+}
+
+// When used in a blockchain context, this block definition is
+// equivalent to the those parts of the block that are relevant to
+// consensus. When used outside the blockchain context, the
+// interpretation of this data structure is context dependent.
+message ConsensusBlock {
+  string id = 1;
+  string previous_id = 2;
+  uint32 index = 3;
+  bytes consensus = 4;
+}
+
+// P2P Messaging
+
+// Send a consensus message to a specific, connected peer
+message ConsensusSendToRequest {
+  ConsensusMessage message = 1;
+  string peer_id = 2;
+}
+
+message ConsensusSendToResponse {
+  enum Status {
+    STATUS_UNSET = 0;
+    OK = 1;
+    INTERNAL_ERROR = 2;
+    NOT_READY = 3;
+  }
+
+  Status status = 1;
+}
+
+// Broadcast a message to all peers
+message ConsensusBroadcastRequest {
+  ConsensusMessage message = 1;
+}
+
+message ConsensusBroadcastResponse {
+  enum Status {
+    STATUS_UNSET = 0;
+    OK = 1;
+    INTERNAL_ERROR = 2;
+    NOT_READY = 3;
+  }
+
+  Status status = 1;
+}
+
+// Event Handlers
+
+// Called when a new consensus message is received
+message ConsensusOnMessageReceivedRequest {
+  ConsensusMessage message = 1;
+}
+
+message ConsensusOnMessageReceivedResponse {
+  enum Status {
+    STATUS_UNSET = 0;
+    OK = 1;
+    INTERNAL_ERROR = 2;
+  }
+
+  Status status = 1;
+}
+
+// Called when a new block is received and validated
+message ConsensusOnNewBlockReceivedRequest {
+  ConsensusBlock block = 1;
+}
+
+message ConsensusOnNewBlockReceivedResponse {
+  enum Status {
+    STATUS_UNSET = 0;
+    OK = 1;
+    INTERNAL_ERROR = 2;
+  }
+
+  Status status = 1;
+}
+
+// Called when a new peer is added
+message ConsensusOnAddPeerRequest {
+  string peer_id = 1;
+}
+
+message ConsensusOnAddPeerResponse {
+  enum Status {
+    STATUS_UNSET = 0;
+    OK = 1;
+    INTERNAL_ERROR = 2;
+  }
+
+  Status status = 1;
+}
+
+// Called when a peer is dropped
+message ConsensusOnDropPeerRequest {
+  string peer_id = 1;
+}
+
+message ConsensusOnDropPeerResponse {
+  enum Status {
+    STATUS_UNSET = 0;
+    OK = 1;
+    INTERNAL_ERROR = 2;
+  }
+
+  Status status = 1;
+}
+
+// Block Creation
+
+// Initialize a new block based on the current chain head and start
+// adding batches to it.
+message ConsensusInitializeBlockRequest {
+
```

Comment:
We may want to add head_id to this later on, to allow the consensus engine to build blocks on top of multiple forks at once or as a check to make sure we are building on top of the block we think we are building on top of.